Multisig/UI/Safe Management/SafeAddressForm/SafeAddressField.swift

```diff
@@ -0,0 +1,152 @@
+    // true - valid
+    // false - invalid
+    var isValid: Binding<Bool?>
+
+    // true - shows progress indicator
+    // false - shows "more" button
+    var isValidating: Binding<Bool>
+
+    // when not empty, shows error under the text
+    var error: Binding<String>
+
+    var isEmpty: Bool { displayText.isEmpty }
+
+    @State
+    private var showsSelector: Bool = false
+
+    var body: some View {
+        VStack(spacing: 15) {
+            VStack(alignment: .leading) {
+                inputView
+                errorView
+            }
+            if isEmpty {
+                CreateSafePrompt()
+            }
+        }
+    }
+
+    // the properties are wrapped into "Group" in order to
+    // conform to the opaque type requirement (always return the same
+    // concrete type)
+    var errorView: some View {
+        Group {
+            if !error.wrappedValue.isEmpty {
+                Text(error.wrappedValue)
+                    .font(Font.gnoBody.weight(.medium))
+                    .foregroundColor(Color.gnoTomato)
+                    .padding(.leading)
+            }
+        }
+    }
+
+    var inputView: some View {
+        Button(action: { self.showsSelector.toggle() }) {
+            HStack {
+                contentView
+
+                Spacer()
+
+                rightView
+
+                AddressInputSelector(isPresented: $showsSelector, text: enteredText)
+            }
+        }
+        .font(Font.gnoBody.weight(.medium))
+        .padding()
+        .frame(height: isEmpty ? 56 : 74)
+        .background(borderView)
+        .foregroundColor(.gnoMediumGrey)
+    }
+
+    var contentView: some View {
+        Group {
+            if isEmpty {
+                Text(title)
+            } else if isAddress.wrappedValue {
+                AddressView(displayText)
+            } else {
+                Text(displayText).foregroundColor(Color.gnoDarkBlue)
+            }
+        }
+    }
+
+    var rightView: some View {
+        Group {
+            if isValidating.wrappedValue {
+                ActivityIndicator(isAnimating: .constant(true), style: .medium)
+            } else if isValid.wrappedValue == true {
+                Image(systemName: "checkmark.circle")
+                    .foregroundColor(.gnoHold)
+            } else if isValid.wrappedValue == false {
+                Image(systemName: "xmark.circle")
+                    .foregroundColor(.gnoTomato)
+            } else {
+                Image(systemName: "ellipsis")
+            }
+        }
+    }
+
+    var borderView: some View {
+        RoundedRectangle(cornerRadius: 10)
+            .stroke(strokeColor, lineWidth: 2)
+    }
+
+    var strokeColor: Color {
+        if let isValid = isValid.wrappedValue {
+            return isValid ? Color.gnoHold50 : Color.gnoTomato
+        } else {
+            return Color.gnoWhitesmoke
+        }
+    }
+
+}
+
+struct SafeAddressInput_Previews: PreviewProvider {
```

Comment:
can you use DEBUG macros not to put it into the production code?

Reply:
hm... why does Apple use it like this?

Reply:
This is annoying, I would not do this if it is not necessary (will have to do this for every view)

Reply:
We need to research more on this. I saw a lot of examples where people do it. But probably it is not necessary and it will be removed by a compiler anyways.